language/vm/vm_runtime/src/interpreter.rs

```diff
@@ -0,0 +1,1208 @@
+    // context are well defined. Obviously certain opcodes require a given context, but
+    // we may be doing better here...
+    pub fn new(
+        module_cache: P,
+        txn_data: TransactionMetadata,
+        data_view: TransactionDataCache<'txn>,
+    ) -> Self {
+        Interpreter {
+            operand_stack: Stack::new(),
+            call_stack: CallStack::new(),
+            gas_meter: GasMeter::new(txn_data.max_gas_amount()),
+            txn_data,
+            event_data: vec![],
+            data_view,
+            module_cache,
+            phantom: PhantomData,
+        }
+    }
+
+    //
+    // The functions below should be reviewed once we clean up the Loader and the
+    // transaction flow. It's not clear whether they are leaking internal of the Interpreter
+    // that would be better exposed via a more proper API.
+    //
+
+    /// Returns the module cache for this interpreter.
+    pub fn module_cache(&self) -> &P {
+        &self.module_cache
+    }
+
+    /// Disables metering of gas.
+    pub fn disable_metering(&mut self) {
+        self.gas_meter.disable_metering();
+    }
+
+    /// Re-enables metering of gas.
+    pub(crate) fn enable_metering(&mut self) {
+        self.gas_meter.enable_metering();
+    }
+
+    /// Returns the gas used by an execution in the `Interpreter`.
+    pub(crate) fn gas_used(&self) -> u64 {
+        self.txn_data
+            .max_gas_amount
+            .sub(self.gas_meter.remaining_gas())
+            .mul(self.txn_data.gas_unit_price)
+            .get()
+    }
+
+    // This is used by the genesis tool and must be deleted and re-worked
+    pub(crate) fn swap_sender(&mut self, address: AccountAddress) -> AccountAddress {
+        let old_sender = self.txn_data.sender;
+        self.txn_data.sender = address;
+        old_sender
+    }
+
+    /// Clear all the writes local to this execution.
+    pub(crate) fn clear(&mut self) {
+        self.data_view.clear();
+        self.event_data.clear();
+    }
+
+    /// Return the list of events emitted during execution.
+    pub(crate) fn get_events(&self) -> Vec<ContractEvent> {
+        self.event_data.clone()
+    }
+
+    /// Generate a `WriteSet` as a result of an execution.
+    pub(crate) fn make_write_set(
+        &mut self,
+        to_be_published_modules: Vec<(ModuleId, Vec<u8>)>,
+    ) -> VMResult<WriteSet> {
+        self.data_view.make_write_set(to_be_published_modules)
+    }
+
+    /// Execute a function.
+    /// `module` is an identifier for the name the module is stored in. `function_name` is the name
+    /// of the function. If such function is found, the VM will execute this function with arguments
+    /// `args`. The return value will be placed on the top of the value stack and abort if an error
+    /// occurs.
+    // REVIEW: this should probably disappear or at the very least only one between
+    // `execute_function` and `interpeter_entrypoint` should exist. It's a bit messy at
```

Comment:
IMHO we would probably need to keep this function. Another use case I could think of is for the block prologue transaction. I would imagine its semantics to be a call to a special private function on chain to reset the block metadata resource.

Reply:
maybe we will need it. The point here is that resolution of the function could happen on the call site and the entry being `interpeter_entrypoint` or vice-versa. Maybe we will need both, not sure.
Once we move forward the cleanup I think we will know better